mods/d2k/maps/shellmap/d2k-shellmap.lua

```diff
@@ -0,0 +1,534 @@
+AtreidesProduceInfantry = function()
+	if HoldAtrProduction then
+		Trigger.AfterDelay(DateTime.Minutes(1), AtreidesProduceInfantry)
+		return
+	end
+
+	local delay = Utils.RandomInteger(AttackDelay[1], AttackDelay[2])
+	local toBuild = { Utils.Random(AtreidesInfantryTypes) }
+	atreides.Build(toBuild, function(unit)
+		IdlingUnitsOfAtr[#IdlingUnitsOfAtr + 1] = unit[1]
+		Trigger.AfterDelay(delay, AtreidesProduceInfantry)
+
+		if #IdlingUnitsOfAtr >= (AttackGroupSize[1] * 2.5) then
+			SendAttackAtr()
+		end
+	end)
+end
+
+AtreidesProduceVehicles = function()
```

Comment:
There is a lot of redundancy in your code. You can simplify it to

```lua
Produce(atreides, infantry)
```

with code like this

```lua
Produce = function(house, units)
	if HoldProduction[house] then
		Trigger.AfterDelay(DateTime.Minutes(1), units)
		return
	end

	local delay = Utils.RandomInteger(AttackDelay[1], AttackDelay[2])
	local toBuild = { Utils.Random(units) }
	atreides.Build(toBuild, function(unit)
		IdlingUnits[house][#IdlingUnits[house] + 1] = unit[1]
		Trigger.AfterDelay(delay, function() Produce(atreides, infantry) end)

		if #IdlingUnits[house] >= (AttackGroupSize[1] * 2.5) then
			SendAttack(house)
		end
	end)
end
```

Note: this isn't tested and needs changes elsewhere in the code as well. Hope you get the idea.